File: LeetCode/C/155-design-最小栈.cpp
/*
 * @lc app=leetcode.cn id=155 lang=cpp
 *
 * [155] 最小栈
 */
#include "main.h"
// @lc code=start
class MinStack {
  private:
    stack<int> m_stack;
    int m_min;

  public:
    /** initialize your data structure here. */
    MinStack() { m_min = INT_MAX; }

    void push(int x) {
        if (x <= m_min) {
            m_stack.emplace(m_min);
            m_min = x;
        }
        m_stack.emplace(x);
    }

    void pop() {
        int t = m_stack.top();
        m_stack.pop();
        if (t == m_min) {
            m_min = m_stack.top();
            m_stack.pop();
        }
    }

    int top() { return m_stack.top(); }

    int getMin() { return m_min; }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(x);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */
// @lc code=end

File: LeetCode/Go/155-design-最小栈.go
/*
 * @lc app=leetcode.cn id=155 lang=golang
 *
 * [155] 最小栈
 */
package leetcode

import "math"

// @lc code=start
type MinStack struct {
	Stack []int
	Min   int
}

/** initialize your data structure here. */
func Constructor() MinStack {
	stack := MinStack{}
	stack.Min = math.MaxInt64
	return stack
}

func (this *MinStack) Push(x int) {
	if x <= this.Min {
		this.Stack = append(this.Stack, this.Min)
		this.Min = x
	}
	this.Stack = append(this.Stack, x)
}

func (this *MinStack) Pop() {
	var top = this.Top()
	this.Stack = this.Stack[:len(this.Stack)-1]
	if top == this.Min {
		this.Min = this.Top()
		this.Stack = this.Stack[:len(this.Stack)-1]
	}
}

func (this *MinStack) Top() int {
	return this.Stack[len(this.Stack)-1]
}

func (this *MinStack) GetMin() int {
	return this.Min
}

/**
 * Your MinStack object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(x);
 * obj.Pop();
 * param_3 := obj.Top();
 * param_4 := obj.GetMin();
 */
// @lc code=end
